### PR TITLE
#register new tenant

### PR DESCRIPTION
**Pull request by AlixH**

### CODE REVIEW

File: src/utils/Utils.tsx [hunk 1]
```diff
@@ -510,7 +510,7 @@ export default class Utils {
     return new Promise((resolve) => setTimeout(resolve, millis));
   }
 
-  public static getParamFromNavigation(route: any, name: string, defaultValue: string, removeValue = false): string | number | boolean | object {
+  public static getParamFromNavigation(route: any, name: string, defaultValue: any, removeValue = false): string | number | boolean | object {
```
Comment (jerome-benoit) [inline src/utils/Utils.tsx:513]: Please consider any type usage as forbidden.

File: src/screens/auth/login/Login.tsx [hunk 2]
```diff
@@ -102,6 +102,10 @@ export default class Login extends BaseScreen<Props, State> {
     let tenantLogo: string;
     // Get tenants
     this.tenants = await this.centralServerProvider.getTenants();
+    if(this.tenants?.length === 0){
```
Comment (jerome-benoit) [inline src/screens/auth/login/Login.tsx:105]: Please use helper Utils.isEmptyArray() to safely test if an array is empty.

File: src/screens/auth/login/Login.tsx [hunk 3]
```diff
@@ -155,6 +160,26 @@ export default class Login extends BaseScreen<Props, State> {
...
+    Alert.alert(
+      I18n.t('authentication.noTenantFoundTitle'),
+      I18n.t('authentication.noTenantFoundMessage'),
+      [
+        {text: I18n.t('authentication.addTenantButton') , onPress:() => {this.goToTenants(true)}},
+        {text: I18n.t('general.close'), style: 'cancel' }
+      ]
+    )
+
+  }
```
Comment (jerome-benoit) [inline src/screens/auth/login/Login.tsx:173]: A blank line is missing between methods.

File: src/screens/tenants/Tenants.tsx [hunk 4]
```diff
@@ -25,6 +25,9 @@ interface State {
   createQrCodeTenantVisible?: boolean
 }
 
+
+
+
```
Comment (jerome-benoit) [inline src/screens/tenants/Tenants.tsx:30]: Unneeded blank lines.

File: src/screens/tenants/Tenants.tsx [hunk 5]
```diff
@@ -41,6 +44,8 @@ export default class Tenants extends BaseScreen<Props, State> {
     await super.componentDidMount();
     const tenants = await this.centralServerProvider.getTenants();
     this.setState({ tenants });
+    const scanQRCode = Utils.getParamFromNavigation(this.props.route, 'openQRCode', this.state.createQrCodeTenantVisible);
+    this.setState({createQrCodeTenantVisible: scanQRCode})
```
Comment (jerome-benoit) [inline src/screens/tenants/Tenants.tsx:48]: Please try to keep the coding style consistent with the style already used ({ value }) across your changes.

**Review by jerome-benoit — CHANGES_REQUESTED** [review]

File: src/screens/tenants/Tenants.tsx [hunk 6]
```diff
@@ -41,6 +41,8 @@ export default class Tenants extends BaseScreen<Props, State> {
     await super.componentDidMount();
     const tenants = await this.centralServerProvider.getTenants();
     this.setState({ tenants });
+    const createQrCodeTenantVisible = Utils.getParamFromNavigation(this.props.route, 'openQRCode', this.state.createQrCodeTenantVisible);
+    this.setState({createQrCodeTenantVisible})
```
Comment (LucasBrazi06) [inline src/screens/tenants/Tenants.tsx:45]: No double setState() here, use only one + missing semicolon.

```
this.setState({
  tenants,
  createQrCodeTenantVisible
});
```

File: src/screens/auth/login/Login.tsx [hunk 7]
```diff
@@ -102,6 +102,10 @@ export default class Login extends BaseScreen<Props, State> {
     let tenantLogo: string;
     // Get tenants
     this.tenants = await this.centralServerProvider.getTenants();
+    if(Utils.isEmptyArray(this.tenants)){
```
Comment (LucasBrazi06) [inline src/screens/auth/login/Login.tsx:105]: Add space after the 'if' statement and before the curly brace

File: src/screens/auth/login/Login.tsx [hunk 8]
```diff
@@ -155,6 +160,26 @@ export default class Login extends BaseScreen<Props, State> {
     }
   }
 
+  private noTenantFoundMessage(){
+    Alert.alert(
+      I18n.t('authentication.noTenantFoundTitle'),
+      I18n.t('authentication.noTenantFoundMessage'),
+      [
+        {text: I18n.t('authentication.addTenantButton') , onPress:() => {this.goToTenants(true)}},
```
Comment (LucasBrazi06) [inline src/screens/auth/login/Login.tsx:168]: Add space after all opened curly braces and before closed ones.
Example:
`{ this.goToTenants(true) }`

File: src/screens/auth/login/Login.tsx [hunk 9]
```diff
@@ -155,6 +160,26 @@ export default class Login extends BaseScreen<Props, State> {
...
+      ]
+    )
+  }
+
+  private goToTenants(openQRCode = false) {
+    this.props.navigation.navigate(
+      'Tenants',
+      {
+        key: `${Utils.randomNumber()}`,
+        openQRCode
```
Comment (LucasBrazi06) [inline src/screens/auth/login/Login.tsx:179]: No need to pass a boolean I think, You can jump right away to the QR-Code screen via the Tenant one.
Check as an example of navigation screen in screen in NotificationManager:
```
      this.navigator.dispatch(
          StackActions.replace(
            'AppDrawerNavigator',
            {
              screen: 'TransactionHistoryNavigator',
              initial: false,
              params: {
                screen: 'TransactionDetailsTabs',
                key: `${Utils.randomNumber()}`,
                params: {
                  params: {
                    transactionID: Utils.convertToInt(notification.data.transactionId)
                  }
                }
              }
            }
          ),
        );

```

File: src/screens/auth/login/Login.tsx [hunk 10]
```diff
@@ -327,117 +352,121 @@ export default class Login extends BaseScreen<Props, State> {
     const { tenantLogo, eula, loading, initialLoading, hidePassword } = this.state;
     // Render
     return initialLoading ? (
-      <Spinner style={formStyle.spinner} color='grey' />
+      <Spinner style={formStyle.spinner} color='grey'/>
     ) : (
-        <View style={style.container}>
```
Comment (LucasBrazi06) [inline src/screens/auth/login/Login.tsx:332]: Be careful with reformatting, it generates a lot of changes in the diff tool with no update.
This one could have been kept in one line so not needed.
You can revert.

File: src/screens/tenants/Tenants.tsx (hunk 6 above)
Comment (LucasBrazi06) [reply src/screens/tenants/Tenants.tsx:45]: But boolean intermediate navigation not needed if you can navigate right away in the QR-Code screen

File: src/screens/tenants/Tenants.tsx [hunk 11]
```diff
@@ -59,6 +61,28 @@ export default class Tenants extends BaseScreen<Props, State> {
     )
   }
 
+  private goToSignUp(tenant : TenantConnection) {
```
Comment (LucasBrazi06) [inline src/screens/tenants/Tenants.tsx:64]: No need a method.

File: src/screens/tenants/Tenants.tsx [hunk 12]
```diff
@@ -77,6 +101,9 @@ export default class Tenants extends BaseScreen<Props, State> {
           tenants,
         });
         Message.showSuccess(I18n.t('general.createTenantSuccess', { tenantName: newTenant.name }));
+
```
Comment (LucasBrazi06) [inline src/screens/tenants/Tenants.tsx:104]: Remove blank lines

File: src/screens/tenants/Tenants.tsx [hunk 13]
```diff
@@ -77,6 +101,9 @@ export default class Tenants extends BaseScreen<Props, State> {
...
+        this.registerToNewTenantMessage(newTenant);
```
Comment (LucasBrazi06) [inline src/screens/tenants/Tenants.tsx:105]: No need a method.

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]
Check my comments ;-)

File: src/screens/auth/login/Login.tsx [hunk 14]
```diff
@@ -155,6 +160,26 @@ export default class Login extends BaseScreen<Props, State> {
...
+      I18n.t('authentication.noTenantFoundTitle'),
+      I18n.t('authentication.noTenantFoundMessage'),
+      [
+        {text: I18n.t('authentication.addTenantButton') , onPress:() => {this.goToTenants(true)}},
+        {text: I18n.t('general.close'), style: 'cancel' }
+      ]
+    )
+  }
+
+  private goToTenants(openQRCode = false) {
```
[the remark this answers is not in the document]
Comment (AlixH) [reply src/screens/auth/login/Login.tsx:174]: Even if I use it twice ?

File: src/screens/auth/login/Login.tsx (hunk 9 above)
Comment (AlixH) [reply src/screens/auth/login/Login.tsx:179]: I don't think it is possible for this, the opening is already handled with a boolean inside the Tenant screen

File: src/screens/auth/login/Login.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (LucasBrazi06) [reply src/screens/auth/login/Login.tsx:168]: Yes you're right, I didn't notice ;-).
My bad.

File: src/screens/auth/login/Login.tsx (hunk 9 above)
Comment (LucasBrazi06) [reply src/screens/auth/login/Login.tsx:173]: Yes you're right, this is part of the Tenant's list.

**Comment by LucasBrazi06** [comment]
@AlixH Don't forget to put back the issue in the To Be Reviewed pipeline when you will have finished

File: src/I18n/languages/en.json [hunk 15]
```diff
@@ -70,6 +70,9 @@
     "reportError": "Report Error"
   },
   "authentication": {
+    "noTenantFoundTitle" : "No tenant found",
```
Comment (LucasBrazi06) [inline src/I18n/languages/en.json:73]: Tenant is called Organization in the UI ;)

File: src/I18n/languages/en.json [hunk 16]
```diff
@@ -70,6 +70,9 @@
...
+    "addTenantButton" : "Scan QR code",
+    "noTenantFoundMessage" : "Do you want to add one?",
```
Comment (LucasBrazi06) [inline src/I18n/languages/en.json:75]: -> Do you want to register a new organization using a QR-Code?

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]

**Comment by LucasBrazi06** [comment]
@AlixH You still not provided translations in other i18n files (it.json...)
[Golden rule 4](https://github.com/LucasBrazi06/ev-dashboard/wiki/Development-Guidelines)

File: src/I18n/languages/fr.json [hunk 17]
```diff
@@ -70,6 +71,8 @@
     "reportError": "Signaler Erreur"
   },
   "authentication": {
+    "noTenantFoundTitle" : "Aucune organisation trouvée",
+    "noTenantFoundMessage" : "Voulez-vous en ajouter une avec un QR-Code?",
```
Comment (LucasBrazi06) [inline src/I18n/languages/fr.json:75]: Not translated from english

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]
Pay attention ;-)

File: src/I18n/languages/fr.json (hunk 17 above)
Comment (AlixH) [reply src/I18n/languages/fr.json]: You mean the line 75 ?

Comment (LucasBrazi06) [reply src/I18n/languages/fr.json]: More like this:
Voulez-vous enregistrer une organisation en utilisant le QR-Code?

File: src/I18n/languages/de.json [hunk 18]
```diff
@@ -70,6 +71,8 @@
     "reportError": "Fehler Melden"
   },
   "authentication": {
+    "noTenantFoundTitle" : "No organization found",
+    "noTenantFoundMessage" : "Do you want to add a new organization using a QR-Code?",
```
Comment (LucasBrazi06) [inline src/I18n/languages/de.json:75]: Almost but in my last comment it was:
Do you want to register a new organization using a QR-Code?

image

Comment (AlixH) [reply src/I18n/languages/de.json]: Ok I was not sure about the true meaning or "register", it is fixed

Comment (AlixH) [reply src/I18n/languages/de.json]: Fixed

File: src/I18n/languages/fr.json (hunk 17 above)
Comment (AlixH) [reply src/I18n/languages/fr.json]: Fixed

**Review by LucasBrazi06 — APPROVED** [review]

**Comment by LucasBrazi06** [comment]
After testing, fix the following issues:
- [ ] After successful tenant scan, message to register is displayed: propose instead of 'Close' and 'Yes' buttons, 'Sign In' and 'Sign Up'
  - When Sign In clicked: navigate to the login screen and prefill the tenant name
  - When Sign Up clicked: navigate to the Sign Up page (current behavior)
- [ ] In Sign Up page, if I click on 'Sign In' to go back to the login screen without registering, the tenant in the login screen should have the newly created tenant set
- [ ] When latest tenant is deleted in the list and you go back to login screen, both input fields user and password are still filled 
(they should be empty)
- [ ] Test, test, test... ;-)

**Pull request merged**